math/src/fft/fft_inputs.rs

```diff
@@ -69,17 +104,98 @@ pub trait FftInputs<B: StarkField> {
     ///
     /// # Panics
     /// Panics if length of the `twiddles` parameter is not self.len() / 2.
-    fn fft_in_place(&mut self, twiddles: &[B]) {
+    fn fft_in_place(&mut self, twiddles: &[E::BaseField]) {
         fft_in_place(self, twiddles, 1, 1, 0);
     }
+
+    // CONCURRENT METHODS
+    // --------------------------------------------------------------------------------------------
+
+    // PERMUTATIONS
+    // ================================================================================================
+
+    #[cfg(feature = "concurrent")]
+    /// Permutes the elements in this input using the permutation defined by the given
+    /// permutation index in a concurrent manner.
+    fn permute_concurrent(&mut self)
+    where
+        Self: Send,
+    {
+        let n = self.len();
+        let num_batches = rayon::current_num_threads().next_power_of_two();
+        let batch_size = n / num_batches.min(n);
+        rayon::scope(|s| {
+            for batch_idx in 0..num_batches {
+                // create another mutable reference to the slice of values to use in a new thread; this
+                // is OK because we never write the same positions in the slice from different threads
+                let values = unsafe { &mut *(&mut *self as *mut Self) };
+                s.spawn(move |_| {
+                    let batch_start = batch_idx * batch_size;
+                    let batch_end = batch_start + batch_size;
+                    for i in batch_start..batch_end {
+                        let j = permute_index(n, i);
+                        if j > i {
+                            values.swap(i, j);
+                        }
+                    }
+                });
+            }
+        });
+    }
+
+    // SPLIT-RADIX FFT
+    // ================================================================================================
+
+    #[cfg(feature = "concurrent")]
+    /// In-place recursive FFT with permuted output.
+    /// Adapted from: https://github.com/0xProject/OpenZKP/tree/master/algebra/primefield/src/fft
```

Comment:
We might want to use a permalink reference, instead. If we always point to `master`, this link can eventually be broken. Also, preferably, it will be good to have a permalink pointing to a source code lines range, instead of a directory. This way, the reference will be unambiguous.